clarity/src/vm/types/serialization.rs

```diff
@@ -882,18 +1145,123 @@ impl Write for WriteCounter {
         to_hex(byte_serialization.as_slice())
     }
-}
 
-impl ClarityDeserializable<Value> for Value {
-    fn deserialize(hex: &str) -> Self {
-        Value::try_deserialize_hex_untyped(hex).expect("ERROR: Failed to parse Clarity hex string")
+    /// Sanitize `value` against pre-2.4 serialization
+    ///
+    /// Returns Some if the sanitization is successful, or was not necessary.
+    /// Returns None if the sanitization failed.
+    ///
+    /// Returns the sanitized value _and_ whether or not sanitization was required.
+    pub fn sanitize_value(
+        epoch: &StacksEpochId,
+        expected: &TypeSignature,
+        value: Value,
+    ) -> Option<(Value, bool)> {
+        // in epochs before 2.4, perform no sanitization
+        if !epoch.value_sanitizing() {
+            return Some((value, false));
+        }
+        let (output, did_sanitize) = match value {
+            Value::Sequence(SequenceData::List(l)) => {
+                let lt = match expected {
+                    TypeSignature::SequenceType(SequenceSubtype::ListType(lt)) => lt,
+                    _ => return None,
+                };
+                if l.len() > lt.get_max_len() {
+                    return None;
+                }
+                let mut sanitized_items = vec![];
+                let mut did_sanitize_children = false;
+                for item in l.data.into_iter() {
+                    let (sanitized_item, did_sanitize) =
+                        Self::sanitize_value(epoch, lt.get_list_item_type(), item)?;
+                    sanitized_items.push(sanitized_item);
+                    did_sanitize_children = did_sanitize_children || did_sanitize;
+                }
+                // do not sanitize list before construction here, because we're already sanitizing
+                let output_list = Value::cons_list_unsanitized(sanitized_items).ok()?;
+                (output_list, did_sanitize_children)
+            }
+            Value::Tuple(tuple_data) => {
+                let tt = match expected {
+                    TypeSignature::TupleType(tt) => tt,
+                    _ => return None,
+                };
+                let mut sanitized_tuple_entries = vec![];
+                let tuple_data_len = tuple_data.len();
+                let mut tuple_data_map = tuple_data.data_map;
+                let mut did_sanitize_children = false;
+                for (key, expect_key_type) in tt.get_type_map().iter() {
+                    let field_data = tuple_data_map.remove(key)?;
+                    let (sanitized_field, did_sanitize) =
+                        Self::sanitize_value(epoch, expect_key_type, field_data)?;
+                    sanitized_tuple_entries.push((key.clone(), sanitized_field));
+                    did_sanitize_children = did_sanitize_children || did_sanitize;
+                }
+                let did_sanitize_tuple = did_sanitize_children || (tt.len() != tuple_data_len);
```

Comment:
For the list type, you checked the len earlier (and also returned None). 

Perhaps something analagous to 
```
if l.len() > lt.get_max_len() {
                     return None;
                 }
```

makes sense in the tuple case.